nussl/separation/composite/overlap_add.py

```diff
@@ -0,0 +1,89 @@
+from librosa.core import audio
+from .. import SeparationBase
+from ... import AudioSignal
+import numpy as np
+import warnings
+import matplotlib.pyplot as plt
+
+class OverlapAdd(SeparationBase):
+    def __init__(self, separation_object, window_length=15, find_permutation=False):
+        """[summary]
+
+        Parameters
+        ----------
+        separation_object : SeparationBase
+            Separation object that overlap and add is applied to.
+        window_length : int, optional
+            Window length of overlap/add window, by default 15 seconds.
+            The hop length will be half the window length.
+        find_permutation : bool, optional
+            Whether or not to find the permutation between chunks before combining.
+        """
+        super().__init__(separation_object.audio_signal)
+
+        self.hop_length = window_length / 2
+        self.window_length = window_length
+        self.separation_object = separation_object
+        self.find_permutation = find_permutation 
+
+    def run(self, *args, **kwargs):
+        audio_signal = self.separation_object.audio_signal
+
+        if audio_signal.signal_duration <= self.window_length:
+            estimates = self.separation_object()
+            self.audio = np.stack([e.audio_data for e in estimates], axis=-1)
+            warnings.warn("Signal duration is less than the window length. "
+                          "Running algorithm on entire signal and returning.")
+            return self.audio
+        
+        win_samples = int(self.window_length * audio_signal.sample_rate)
+        hop_samples = int(self.hop_length * audio_signal.sample_rate)
+        audio_signal.zero_pad(hop_samples, hop_samples)
+        num_samples = audio_signal.signal_length
+        
+        audio_data = np.zeros_like(audio_signal.audio_data)
+        chunk_starts = np.arange(0, num_samples - hop_samples, hop_samples)
+        estimates_from_chunks = []
+        
+        for i, start_idx in enumerate(chunk_starts):
```

Comment:
One thing I wanted to do a while ago is make a generator attached to AudioSignal that would do all of this. So you would use it like:

```python
for region in audio_signal.get_next_region('hanning', win_samples):
    # do stuff
```

Something like that. It would do all of this fancy stuff you've done here within the AudioSignal object, so that the logic is usable elsewhere. Do you think that it would be worth moving what you have here? Would it be useful? Then this OverlapAdd isn't so much a class as it is a recipe. I haven't fully though through everything that would change. Thoughts?

Reply:
Feels like an easy refactor in the future but wanted to keep this one simple and restricted to one method for now, if that's alright.